tests/services/data_reduction_test.py

```diff
@@ -0,0 +1,394 @@
+
+
+def make_reduction_app(instrument: str) -> BeamlimeApp:
+    builder = make_reduction_service_builder(instrument=instrument)
+    return BeamlimeApp.from_service_builder(builder)
+
+
+@pytest.mark.parametrize("instrument", available_instruments())
+def test_publishes_workflow_specs_on_startup(instrument: str) -> None:
+    app = make_reduction_app(instrument=instrument)
+    sink = app.sink
+
+    assert len(sink.messages) == 1
+    message = sink.messages[0]
+    assert message.stream == CONFIG_STREAM_ID
+    assert isinstance(message.value, ConfigUpdate)
+    assert isinstance(message.value.config_key, models.ConfigKey)
+    assert isinstance(message.value.value, models.WorkflowSpecs)
+    if instrument in ('bifrost', 'dummy', 'loki'):
+        assert len(message.value.value.workflows) > 0
+
+
+@pytest.mark.parametrize("instrument", ['bifrost', 'dummy'])
+def test_can_configure_and_stop_workflow_with_detector(
+    instrument: str, caplog: pytest.LogCaptureFixture
+) -> None:
+    caplog.set_level(logging.INFO)
+    app = make_reduction_app(instrument=instrument)
+    sink = app.sink
+    service = app.service
+    workflow_specs = sink.messages[0].value.value
+    workflow_name = {'bifrost': 'spectrum-view', 'dummy': 'Total counts'}[instrument]
+    workflow_id, spec = _get_workflow_by_name(workflow_specs, workflow_name)
+    sink.messages.clear()  # Clear the initial message
+
+    # Assume workflow is runnable for all source names
+    config_key = models.ConfigKey(
+        source_name=None, service_name="data_reduction", key="workflow_config"
+    )
+    workflow_config = models.WorkflowConfig(
+        identifier=workflow_id,
+        values={param.name: param.default for param in spec.parameters},
+    )
+    # Trigger workflow start
+    app.publish_config_message(key=config_key, value=workflow_config.model_dump())
+
+    app.publish_events(size=2000, time=2)
+    service.step()
+    assert len(sink.messages) == 1
+    # Events before workflow config was published should not be included
+    assert sink.messages[0].value.values.sum() == 2000
+    service.step()
+    assert len(sink.messages) == 1
+    app.publish_events(size=3000, time=4)
+    service.step()
+    assert len(sink.messages) == 2
+    assert sink.messages[1].value.values.sum() == 5000
+
+    # More events but the same time, should not publish again
+    app.publish_events(size=1000, time=4)
+    service.step()
+    assert len(sink.messages) == 2
+
+    # Later time should publish again, including the previous events with duplicate time
+    app.publish_events(size=1000, time=5)
+    service.step()
+    assert len(sink.messages) == 3
+    assert sink.messages[2].value.values.sum() == 7000
+
+    # Stop workflow
+    app.publish_config_message(key=config_key, value=None)
+    app.publish_events(size=1000, time=10)
+    service.step()
+    app.publish_events(size=1000, time=20)
+    service.step()
+    assert len(sink.messages) == 3
```

Comment:
```suggestion
    assert len(sink.messages) == 3
    service.stop()
```

to clean up properly or not required?

Reply:
And same for tests below.

Reply:
The service is never actually started. The tests call `step` which does one "iteration" of a running service. This removes complications such as having to wait until the service thread (made when starting) did the next iteration, etc.